### PR TITLE
Rdrp 1008 mig ready

### PR DESCRIPTION
**Pull request by AnneONS**
## Pull Request submission

*Insert detailed bullet points about your changes here!*

*Insert any instructions to help the reviewer, e.g. "install new requirements from `requirements.txt`"*

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: .github/workflows/pytest-action.yaml [hunk 1]
```diff
@@ -19,10 +19,10 @@ jobs:
         run: |
           awk '!/pydoop.*/' requirements.txt> temp && mv temp requirements.txt
 # 3) Set up Python
-      - name: Set up Python 3.6
+      - name: Set up Python 3.10
         uses: actions/setup-python@v4
```
Comment (zorge69) [inline .github/workflows/pytest-action.yaml]: I think it should be v5, see https://github.com/actions/setup-python

File: main.py [hunk 2]
```diff
@@ -2,6 +2,9 @@
 import time
 import os
 
+# os.chdir('../../../home/cdsw/research-and-development')
```
Comment (zorge69) [inline main.py]: Passing it for now; we will need to do it as in branch 994.

File: make_dirs_main.py [hunk 3]
```diff
@@ -4,80 +4,88 @@
...
-                    "output_intram_uk_itl1": {},
-                    "output_intram_uk_itl2": {},
-                    "output_long_form": {},
-                    "output_ni_sas": {},
-                    "output_short_form": {},
-                    "output_status_filtered_qa": {},
-                    "output_tau": {},
+    global config
+    root = "/user/anne.griffith/res_dev/2022_surveys"
+    # root = "/bat/res_dev/project_data"
```
Comment (zorge69) [inline make_dirs_main.py]: We should switch to /bat/res_dev/project_data

File: make_dirs_main.py [hunk 4]
```diff
@@ -4,80 +4,88 @@
...
+        },
+        "mappers": {
+            "v1": {},
+        },
+        "PNP": {
+            "01_staging": {},
+        },
     }
+
     dir_list = tree_to_list(tree, prefix=root)
```
Comment (zorge69) [inline make_dirs_main.py]: Bug fix in 994

File: singleton_main.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+'''
```
Comment (zorge69) [inline singleton_main.py]: singleton main no longer needed, removed in boto_anne

File: src/outlier_detection/auto_outliers.py [hunk 6]
```diff
@@ -58,7 +56,7 @@ def filter_valid(df: pd.DataFrame, value_col: str) -> pd.DataFrame:
         pd.DataFrame: The filtered dataframe
     """
     sample_cond = df.selectiontype == "P"
-    status_cond = df.statusencoded.isin(["210", "211"])
+    status_cond = df["status"].isin(["Clear", "Clear - overridden"])
```
Comment (zorge69) [inline src/outlier_detection/auto_outliers.py]: This is unrelated to this feature. I am happy to pass it, but please let's have one feature/one branch/one PR.

File: src/pipeline.py [hunk 7]
```diff
@@ -20,7 +20,7 @@
 from src.estimation.estimation_main import run_estimation
 from src.site_apportionment.site_apportionment_main import run_site_apportionment
 from src.outputs.outputs_main import run_outputs
-
+from src.utils.singleton_boto import SingletonBoto
```
Comment (zorge69) [inline src/pipeline.py]: It needs to be conditional on platform==s3. Changed.

File: src/pipeline.py [hunk 8]
```diff
@@ -48,10 +49,12 @@ def run_pipeline(user_config_path, dev_config_path):
...
 
     if platform == "s3":
-        from src.utils import s3_mods as mods
+        #create singletion boto3 client object & pass in bucket string
+        boto3_client = SingletonBoto.get_client()
 
+        from src.utils import s3_mods as mods
         # Creating boto3 client and adding it to the config dict
-        config["client"] = mods.create_client(config)
+        config["client"] = boto3_client
```
Comment (zorge69) [inline src/pipeline.py]: Not needed; removed in boto_anne

File: src/user_config.yaml [hunk 9]
```diff
@@ -52,8 +52,8 @@ global:
 hdfs_paths:
   root: "/ons/rdbe_dev/"
   # staging input paths
-  snapshot_path: "/ons/rdbe_dev/spp_snapshots/2023_snapshots/snapshot-202312-002-85ae5659-7147-42c3-a5dd-d69beccc9e09.json" # Run until point of freezing - last run is first freeze
-  updated_snapshot_path: "/ons/rdbe_dev/spp_snapshots/2023_snapshots/snapshot-202312-002-b9b6048a-51c9-4669-919a-e92fc6e9c433.json"
+  snapshot_path: "/ons/rdbe_dev/spp_snapshots/2023_snapshots/snapshot-202312-002-85ae5659-7147-42c3-a5dd-d69beccc9e09.json" #  Run until point of freezing - last run is first freeze
+  updated_snapshot_path: "s3a://onscdp-dev-data01-5320d6ca/bat/res_dev/anonymised/snapshot_202312_a_copy_of_202012_anon.json"
```
Comment (zorge69) [inline src/user_config.yaml]: we don't need full paths with s3a prefix. this should be fixed in 994

**Review by zorge69 — APPROVED** [review]

File: main.py (hunk 2 above)
Comment (AnneONS) [reply main.py]: I agree

File: src/outlier_detection/auto_outliers.py (hunk 6 above)
Comment (AnneONS) [reply src/outlier_detection/auto_outliers.py]: sorry! I did it as the unit tests were failing without this in 3.10

Comment (zorge69) [reply src/outlier_detection/auto_outliers.py]: Thanks you, makes sense!
